rust/kernel/spi.rs

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::bindings;
+use crate::c_types;
+use crate::error::{Error, Result};
+use crate::CStr;
+use alloc::boxed::Box;
+use core::pin::Pin;
+
+#[derive(Clone, Copy)]
+pub struct SpiDevice(*mut bindings::spi_device);
+
+impl SpiDevice {
+    pub unsafe fn from_ptr(dev: *mut bindings::spi_device) -> Self {
```

Comment:
I am not sure if I understand the use of `unsafe` here. Or the `// SAFETY:` guarantee in the macros below.

Reply:
This was done because of bjorn3's [comment](https://github.com/Rust-for-Linux/linux/pull/264#discussion_r631385301)

Reply:
> I am not sure if I understand the use of `unsafe` here. Or the `// SAFETY:` guarantee in the macros below.

What bjorn said is that to the user, this function is unsafe as it uses a raw pointer which could be NULL or invalid. If we mark it as unsafe then we show the user that it holds no guarantees. Regarding the SAFETY comment: When our Spi driver gets probed (or removed, or shut down), the `spi_device` pointer is provided by the kernel. It is a pointer to the SPI client probing interacting with our device. Therefore, it is safe to pass to this function as it's just a valid pointer that the kernel gives us

Reply:
> What bjorn said is that to the user, this function is unsafe as it uses a raw pointer which could be NULL or invalid. If we mark it as unsafe then we show the user that it holds no guarantees.

Not exactly -- to clarify a bit: the function does not use the pointer, and so it could potentially be a safe one. The problem is that other functions later on actually use it, and they are not `unsafe`. Since it is best to require that the pointer is valid just once, it is best to do it here, and then `SpiDevice` gains a type invariant, which then `Spi` can use to not need to be `unsafe`.

Speaking of which: you are missing the documentation and the `# Safety` section on the function, etc. The `SpiDevice` should also have documentation, including this type invariant. See other examples in `rust/kernel` (you can also wait a bit, since I have a PR pending with more docs explaining this).